src/Grant/GrantFactory.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace League\OAuth2\Client\Grant;
+
+class GrantFactory
+{
+    /**
+     * @var array
+     */
+    protected $registry = [];
+
+    /**
+     * Define a grant class in the registry.
+     *
+     * @param  string $name
+     * @param  string $class
+     * @return $this
+     */
+    public function setGrant($name, $class)
+    {
+        $this->registry[$name] = $this->checkGrant($class);
+
+        return $this;
+    }
+
+    /**
+     * Get a grant instance by name.
+     *
+     * If the grant has not be registered, a default grant will be loaded.
+     *
+     * @param  string $name
+     * @param  string $options
+     * @return GrantInterface
+     */
+    public function getGrant($name, array $options = [])
+    {
+        if (empty($this->registry[$name])) {
+            $this->registry[$name] = $this->getGrantClass($name);
+        }
+
+        $class = $this->registry[$name];
+
+        return new $class($options);
```

Comment:
This new constructor signature allows for some future possibilities, though the current implementation in `AbstractProvider` doesn't pass anything in right now. I don't feel strongly about this change, and I'm not sure where the `$options` would come from inside of the provider.

In some ways, this feels like a bit of a poor man dependency injection... I wonder if it would be better for the registry to contain *instances* instead of *classes* so that the same grants are reused. Since they are stateless, it would make a lot more sense:

```php
$grant = new MyCustomGrant($options);
$provider->getGrantFactory()->setGrant('custom', $grant);
```

Having instances in the registry would mean that all grant types are singleton, but I don't see a problem with that.